bluetooth: host: HCI support for CS configuration

bugo-nordic opened the pull request:
Adds following HCI support for CS configuration:
- LE CS Create Config command
- LE CS Remove Config command
- LE CS Config Complete event

olivier-le-sage commented on subsys/bluetooth/shell/cs.c:
```diff
@@ -194,6 +194,184 @@ static int cmd_cs_test_simple(const struct shell *sh, size_t argc, char *argv[])
+	params.ch3c_shape = BT_CONN_LE_CS_CONFIG_CH3C_SHAPE_HAT;
+	params.ch3c_jump = 2;
+
+	memset(params.channel_map, 0xFF, 10);
```
I wonder if we should expose a "nice" way of setting channel maps. I also wrote code like:

![channel map](https://github.com/user-attachments/assets/f22ec2d9-afb1-4b11-a17d-6592d5195bfb)

but it's not very readable

bugo-nordic replied:
I can add a function to set all the valid bits in a given chmap

olivier-le-sage replied:
maybe also a macro similar to CS_CHANNEL_BIT() that lets you access channel X in the array

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -413,6 +413,124 @@ struct bt_conn_le_cs_fae_table {
+/** Channel sounding config main mode */
+enum bt_conn_le_cs_config_main_mode {
+	/** Mode-1 (RTT) */
+	BT_CONN_LE_CS_CONFIG_MAIN_MODE_1 = 1,
```
can use BT_HCI_OP_LE_CS_MAIN_MODE_X

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -413,6 +413,124 @@ struct bt_conn_le_cs_fae_table {
+/** Channel sounding config sub mode */
+enum bt_conn_le_cs_config_sub_mode {
+	/** Unused */
+	BT_CONN_LE_CS_CONFIG_SUB_MODE_UNUSED = 0xFF,
```
can use BT_HCI_OP_LE_CS_SUB_MODE_X

olivier-le-sage commented on subsys/bluetooth/shell/cs.c:
```diff
@@ -194,6 +194,184 @@ static int cmd_cs_test_simple(const struct shell *sh, size_t argc, char *argv[])
+				return -ENOEXEC;
+			}
+
+			sys_mem_swap(params.channel_map, 10);
```
swap ?

bugo-nordic replied:
hex2bin generates big endian value so we're swapping here to make it le

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -506,6 +506,51 @@ struct bt_cs_test_param {
+	enum bt_conn_le_cs_config_ch3c_shape ch3c_shape;
+	/** Number of channels skipped in each rising and falling sequence  */
+	uint8_t ch3c_jump;
+	/** Channel map */
```
should probably document which channels aren't allowed and/or RFU

olivier-le-sage replied:
there's also a requirement to provide at least 15 channels

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -413,6 +413,124 @@ struct bt_conn_le_cs_fae_table {
+};
+
+/** Channel sounding config role */
+enum bt_conn_le_cs_config_role {
```
maybe we can rename this to

enum bt_conn_le_cs_role {
BT_CONN_LE_CS_ROLE_INITIATOR = BT_HCI_OP_LE_CS_INITIATOR_ROLE,
BT_CONN_LE_CS_ROLE_REFLECTOR = BT_HCI_OP_LE_CS_REFLECTOR_ROLE,
}

so that we can get rid of enum bt_cs_test_role

I say this assuming it's always 0 = initiator 1 = reflector throughout these commands

bugo-nordic replied:
Yes good shout

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -413,6 +413,130 @@ struct bt_conn_le_cs_fae_table {
 };
 
+/** Channel sounding main mode */
+enum bt_conn_le_cs_main_mode {
```
There seems to be some inconsistency with the API naming. For the public functions you only use "cs", e.g. "bt_cs_create_config()", however here you use "le_cs". I think it would be good either use the "le_" prefix everywhere or nowhere (my preference is the latter). The only exception could be the HCI-related definitions, since there it's better to follow whatever the HCI spec calls the events, commands and associated parameters.

bugo-nordic replied:
You are right. API exposed from `cs.h` is specific to LE CS so I agree with you that we don't need to name it `le_cs` apart from the HCI parts.

Where the inconsistency happens is in `conn.h` which appear to contain both LE and bluetooth classic features. Additional prefix there might be useful to distinguish between the two but I see your point in terms of the inconsistency.

Happy to change the cs related definitions in `conn.h` if you prefer that. One downside to that will be the inconsistency of the naming between some other LE features (`bt_conn_le_subrating*`, `bt_conn_le_tx_power*` etc.) and CS.

jhedberg replied:
I took a look at existing LE-related definitions in `conn.h` and it seems the convention is to have `le_` here, so it's probably best to just follow this. What's also messy, is that the `bt_conn_cb` members (even non-CS related ones) *don't* have `le_` even though their other corresponding APIs do have it. (e.g. LE power reporting or LE Subrating related APIs).

bugo-nordic replied:
Good point with the callbacks. Not sure about the previous features but I lean towards adding le for CS callbacks at least. Just so we can stick to a standard for the newly added codes. In long term we would have more callbacks that follow the correct approach this way.

What do you think?

jhedberg replied:
@jori-nordic @alwa-nordic @carlescufi any thoughts regarding the conventions & consistency with LE-specific Bluetooth APIs?

jhedberg replied:
The Zephyr Bluetooth host stack has supported both BR/EDR and LE since the beginning (2015 or so) and the approach has been that anything LE-specific should have "le" in the name. Since then, especially when it seemed like no-one was using BR/EDR and it might be completely removed, there's been a trend to simplify things and just omit the "LE" part. Now that Bluetooth Classic support has been revived it seems like this might not have been that wise after all.

jori-nordic replied:
Since classic is on its way out, I would err on the side of removing `le_` and instead adding `br_` on APIs.

jhedberg replied:
> Since classic is on its way out, I would err on the side of removing `le_` and instead adding `br_` on APIs.

No objections to this, but to be clear it's only "on the way out" from a Bluetooth SIG long-term plan perspective. In a Zephyr context I'd say it's very much "on the way in", at least based on the ever growing list of pull requests I have to review :)

jhedberg replied:
@lylezhu2012 FYI (see earlier comments)

jori-nordic replied:
haha yes for sure. Any way, having a clear split would be good for new (both bluetooth & zephyr) users.

alwa-nordic replied:
+1 for `bt_le_` for clear division.

bugo-nordic replied:
Will add `le` prefix for CS callbacks if there are no objections then

jhedberg replied:
> Will add `le` prefix for CS callbacks if there are no objections then

Hmm, @jori-nordic said that he'd prefer removing `le_`, to which I said I have no objections. @alwa-nordic said the opposite. I don't think anybody was strongly objecting one way or another, but I didn't get the impression that there was a strong consensus either. What I care about is that we try to move towards API consistency, especially within specific features (like CS) but also within the Bluetooth API in general. As for BR/EDR, I *think* we should be in pretty good shape there, i.e. most of those APIs should already have `br_` in the names.

bugo-nordic replied:
> Any way, having a clear split would be good for new (both bluetooth & zephyr) users.

I interpreted this as clear split with the use of `le` and `br` prefixes but maybe I misunderstood.

To move this forward, do you suggest applying changes, leaving as is or having further discussion?

jori-nordic replied:
Seeing the amount of `le_` around in the API, I would prefer `bt_le_` and `bt_br_`.

jhedberg replied:
Ok. Let's go with that then, so please make sure any public functions have either `bt_conn_le_cs_*` or `bt_le_cs_*` as their name and any `bt_conn_cb` callbacks `le_cs_*`.

bugo-nordic committed:
bluetooth: host: HCI support for CS configuration

Adds following HCI support for CS configuration:
- LE CS Create Config command
- LE CS Remove Config command
- LE CS Config Complete event

Two callbacks have been added to notify the application
when a new CS configuration is created or an existing
CS configuration is removed.

Signed-off-by: Burak Gorduk <burak.gorduk@nordicsemi.no>

bugo-nordic committed:
bluetooth: host: Use consistent API naming for CS

This change adds bt_le_cs prefix to public CS API as well as
adding le_cs prefix to CS related bt_conn callbacks. This is
to ensure consistency within CS API while maintaining the
separation between the Bluetooth LE and Classic features.

Signed-off-by: Burak Gorduk <burak.gorduk@nordicsemi.no>

bugo-nordic commented:
rebased
